In [1]:
import numpy as np

import torch
import torch.nn as nn
import torch.nn.functional as F
import higher

from transformers import GPT2Tokenizer, GPT2LMHeadModel

import utils

## Load Model, Data

In [2]:
device = 'cuda' if torch.cuda.is_available() else 'cpu'

Try using GPT2 vs distilgpt2 (KL div actually goes up though)

In [3]:
def loadOTSModel():
    model = GPT2LMHeadModel.from_pretrained("gpt2")
    tokenizer = GPT2Tokenizer.from_pretrained('gpt2')
    tokenizer.pad_token = tokenizer.eos_token

    return model, tokenizer

In [4]:
model, tokenizer = loadOTSModel()

In [5]:
dataloader = utils.retrieveDataloader(
        tokenizer, 
        bs=1, 
        dataset='train'
    )

In [6]:
len(dataloader)

20892

In [7]:
for train_step, (lm_data, edit_example, _) in enumerate(dataloader):

    lm_tokens, lm_mask = lm_data
    lm_tokens, lm_mask = lm_tokens.to(device), lm_mask.to(device)
    edit_tokens, edit_mask = edit_example
    edit_tokens, edit_mask = edit_tokens.to(device), edit_mask.to(device)

    lm_labels = lm_tokens.masked_fill(lm_mask == 0, -100)
    edit_labels = edit_tokens.masked_fill(edit_mask == 0, -100) 
    break

## Double Check Data

In [8]:
lm_tokens

tensor([[14291,   465, 12928,   422, 29032,  3841,   837, 18322,  4488,   329,
         39964,  1578,   764,  2102,   837,  1708,   281,  5095,   287,   662,
          2488,    12,    31,  1622,   837, 18322,  1043,  6443,  3614,   837,
           878,  4191,  1642,  1478,  4652, 11057,   764,   679,  4191,  1364,
           262,  3430,  1863,   351,  5891, 26318,  4186,  6706,   626,   735,
           287,  3269,  1853,   764,   198, 50256, 50256, 50256, 50256, 50256,
         50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256,
         50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256,
         50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256,
         50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256,
         50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256,
         50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256,
         50256, 50256, 50256, 50256, 50256, 50256, 5

In [9]:
lm_mask

tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0]], device='cuda:0')

In [10]:
lm_labels

tensor([[14291,   465, 12928,   422, 29032,  3841,   837, 18322,  4488,   329,
         39964,  1578,   764,  2102,   837,  1708,   281,  5095,   287,   662,
          2488,    12,    31,  1622,   837, 18322,  1043,  6443,  3614,   837,
           878,  4191,  1642,  1478,  4652, 11057,   764,   679,  4191,  1364,
           262,  3430,  1863,   351,  5891, 26318,  4186,  6706,   626,   735,
           287,  3269,  1853,   764,   198,  -100,  -100,  -100,  -100,  -100,
          -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,
          -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,
          -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,
          -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,
          -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,
          -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,
          -100,  -100,  -100,  -100,  -100,  -100,  

In [11]:
tokenizer.decode(lm_tokens[lm_labels != -100])

'Following his departure from Dartford, Julian signed for Sutton United. However, following an injury in pre @-@ season, Julian found opportunities limited, before eventually making 14 league appearances. He eventually left the club along with fellow goalkeeper Tom Lovelock in January 2015.\n'

In [12]:
tokenizer.decode(lm_labels[lm_labels != -100])

'Following his departure from Dartford, Julian signed for Sutton United. However, following an injury in pre @-@ season, Julian found opportunities limited, before eventually making 14 league appearances. He eventually left the club along with fellow goalkeeper Tom Lovelock in January 2015.\n'

In [13]:
tokenizer.decode(lm_tokens[lm_labels != -100]) == tokenizer.decode(lm_labels[lm_labels != -100])

True

## KL Divergence on same model, same data, model.train()

In [40]:
model.train()
model.to(device)
model_out1 = model(lm_tokens, attention_mask=lm_mask, labels=lm_labels)
model_out2 = model(lm_tokens, attention_mask=lm_mask, labels=lm_labels)

In [41]:
model_out1.loss

tensor(4.7208, device='cuda:0', grad_fn=<NllLossBackward>)

In [42]:
model_out2.loss

tensor(4.8962, device='cuda:0', grad_fn=<NllLossBackward>)

In [43]:
model_out1.logits

tensor([[[ -83.5357,  -81.1782,  -84.8859,  ...,  -87.5771,  -90.1763,
           -82.9961],
         [ -82.6794,  -80.3046,  -84.1308,  ...,  -82.5107,  -85.8973,
           -81.8964],
         [ -69.7387,  -69.7368,  -75.8238,  ...,  -77.5415,  -78.3443,
           -71.2514],
         ...,
         [ -94.0070,  -87.0880,  -89.3705,  ..., -108.8873, -109.0895,
           -94.7492],
         [ -89.5154,  -82.6523,  -84.5890,  ..., -102.4380, -102.1378,
           -89.7528],
         [ -97.9364,  -91.3140,  -93.8039,  ..., -111.7013, -111.7482,
           -97.7645]]], device='cuda:0', grad_fn=<UnsafeViewBackward>)

In [44]:
model_out1.logits - model_out2.logits

tensor([[[ 13.1518,  13.4113,  13.2810,  ...,  12.9238,  13.3144,  11.1464],
         [ 22.9788,  22.9977,  24.8985,  ...,  23.4227,  22.8717,  22.8075],
         [  4.4584,   4.0121,   3.3618,  ...,   3.2623,   2.7055,   3.0508],
         ...,
         [  9.9888,  11.1706,  10.8695,  ...,   7.3800,   7.8151,   8.8785],
         [  7.8818,   7.3659,   7.7996,  ...,   7.7815,   7.9289,   8.0477],
         [-14.1567, -13.9872, -14.4588,  ..., -12.6759, -11.4402, -17.9914]]],
       device='cuda:0', grad_fn=<SubBackward0>)

In [45]:
model_out1.logits.shape

torch.Size([1, 200, 50257])

This seems like the wrong dimension

In [70]:
torch.sum(F.softmax(model_out1.logits, dim=1), dim=1)

tensor([[1., 1., 1.,  ..., 1., 1., 1.]], device='cuda:0',
       grad_fn=<SumBackward1>)

In [72]:
torch.sum(F.softmax(model_out1.logits, dim=1), dim=1).shape

torch.Size([1, 50257])

In [75]:
F.kl_div(
    F.log_softmax(model_out1.logits, dim=1),
    F.log_softmax(model_out2.logits, dim=1),
    reduction='batchmean',
    log_target=True
)

tensor(4.4246e-14, device='cuda:0', grad_fn=<DivBackward0>)

This seems like the right dimension - softmax over the 50257 words in vocab for each position in sequence 1-200

In [73]:
torch.sum(F.softmax(model_out1.logits, dim=-1), dim=-1)

tensor([[1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000,
         1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000,
         1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000,
         1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000,
         1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000,
         1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000,
         1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000,
         1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000,
         1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000,
         1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000,
         1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000,
         1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000,
         1.0000, 1.0000, 1.0

In [74]:
torch.sum(F.softmax(model_out1.logits, dim=-1), dim=-1).shape

torch.Size([1, 200])

In [48]:
F.kl_div(
    F.log_softmax(model_out1.logits, dim=-1),
    F.log_softmax(model_out2.logits, dim=-1),
    reduction='batchmean',
    log_target=True
)

tensor(73.3443, device='cuda:0', grad_fn=<DivBackward0>)

In [69]:
kl_loss = nn.KLDivLoss(reduction = 'batchmean')
l_loc = kl_loss(
    F.log_softmax(model_out1.logits, dim=-1),
    F.softmax(model_out2.logits, dim=-1)
)
l_loc

tensor(73.3443, device='cuda:0', grad_fn=<DivBackward0>)

## KL Divergence on same model, same data, model.eval()

In [67]:
model.eval()
model_eval1 = model(lm_tokens, attention_mask=lm_mask, labels=lm_labels)
model_eval2 = model(lm_tokens, attention_mask=lm_mask, labels=lm_labels)
F.kl_div(
    F.log_softmax(model_eval1.logits, dim=-1),
    F.log_softmax(model_eval2.logits, dim=-1),
    reduction='batchmean',
    log_target=True
)

tensor(0., device='cuda:0', grad_fn=<DivBackward0>)

Large KL divergence coming from dropout?

## From editable code:

In [ ]:
def edit(self, inputs, targets, max_steps=None, model_kwargs=None, loss_kwargs=None, opt_kwargs=None, **kwargs):
    """
    Attempts to edit model (out-of-place) and return an edited copy
    :param inputs: data that is fed into the model
    :param targets: reference answers that are fed into loss function
    :param max_steps: after this many gradient steps the process is terminated
    :param model_kwargs: optional extra model inputs, used as model(inputs, **model_params)
    :param loss_kwargs: optional extra loss parameters, self.loss_function(model(inputs), targets, **loss_params)
    :param opt_kwargs: optional overrides for optimizer.get_initial_state
    :param kwargs: extra parameters passed to optimizer.step
    :returns: edited_model, is_edit_successful, final_loss, gradients_steps
    :rtype: Editable.EditResult
    """
    model_kwargs, loss_kwargs, opt_kwargs = model_kwargs or {}, loss_kwargs or {}, opt_kwargs or {}
    optimizer_state = self.optimizer.get_initial_state(self, **opt_kwargs)
    editable = self

    for step in count():
        prediction = editable(inputs, **model_kwargs)
        loss = self.loss_function(prediction, targets, **loss_kwargs)

        if self.is_edit_finished(**locals()):
            return self.EditResult(editable, success=True, loss=loss, complexity=step)
        elif step >= (max_steps or self.max_steps):
            return self.EditResult(editable, success=False, loss=loss, complexity=step)

        optimizer_state, editable = self.optimizer.step(
            optimizer_state, editable, loss, parameters=editable.get_editable_parameters(editable.module), **kwargs)

In [66]:
def train_on_batch(self, x_batch, y_batch, x_edit, y_edit, prefix='train/', is_train=True, **kwargs):
    """ Performs a single gradient update and reports metrics """
    x_batch, y_batch = map(torch.as_tensor, (x_batch, y_batch))
    self.opt.zero_grad()

    with training_mode(self.model, is_train=is_train):
        logits = self.model(x_batch)

    main_loss = self.loss_function(logits, y_batch).mean()

    with training_mode(self.model, is_train=False):
        model_edited, success, editability_loss, complexity = self.model.edit(x_edit, y_edit, **kwargs)
        logits_updated = model_edited(x_batch)

    stability_loss = - (F.softmax(logits.detach(), dim=1) * F.log_softmax(logits_updated, dim=1)).sum(dim=1).mean()

    final_loss = main_loss + self.stability_coeff * stability_loss + self.editability_coeff * editability_loss

    metrics = dict(
        final_loss=final_loss.item(), stability_loss=stability_loss.item(),
        editability_loss=editability_loss.item(), main_loss=main_loss.item(),
    )

    final_loss.backward()

    if self.max_norm is not None:
        metrics['grad_norm'] = torch.nn.utils.clip_grad_norm_(self.model.parameters(), max_norm=self.max_norm)
    self.opt.step()

    return self.record(**metrics, prefix=prefix)

Note stability loss:

In [64]:
-(F.softmax(model_out2.logits.detach(), dim=1) * F.log_softmax(model_out1.logits, dim=1)).sum(dim=1).mean()

tensor(1.4916e-18, device='cuda:0', grad_fn=<NegBackward>)

Might need to change dimensions though:

In [65]:
-(F.softmax(model_out2.logits.detach(), dim=-1) * F.log_softmax(model_out1.logits, dim=-1)).sum(dim=-1).mean()

tensor(6.2890, device='cuda:0', grad_fn=<NegBackward>)

In [62]:
F.log_softmax(model_out1.logits, dim=1).shape

torch.Size([1, 200, 50257])

Conclusions:
* Use different stability loss
* Double check on softmax dimensions in KL div
* Take original model logits in training mode
* Performs edit function on model in eval mode (`with training_mode(self.model, is_train=False):`)